File: src/main/java/org/cubexmc/metro/service/strategy/GlobalDispatchStrategy.java
package org.cubexmc.metro.service.strategy;

import java.util.List;

import org.cubexmc.metro.Metro;
import org.cubexmc.metro.model.Line;
import org.cubexmc.metro.service.DispatchStrategy;
import org.cubexmc.metro.service.LineService;
import org.cubexmc.metro.train.TrainInstance;

public class GlobalDispatchStrategy implements DispatchStrategy {
    @Override
    public void tick(LineService service, long currentTick) {
        if (!service.isDepartureWindow(currentTick)) {
            return;
        }

        if (service.isLoopLine() && isLoopTrainArrivingSoon(service, currentTick)) {
            return;
        }

        service.markDeparture(currentTick);
        service.spawnTrain(currentTick);
    }

    private boolean isLoopTrainArrivingSoon(LineService service, long currentTick) {
        if (!service.isLoopLine()) {
            return false;
        }
        Line line = service.getLine();
        if (line == null) return false;
        List<String> ordered = line.getOrderedStopIds();
        if (ordered.isEmpty()) return false;
        String startStopId = ordered.get(0);
        double threshold = Math.max(2.0, service.getHeadwaySeconds() * 0.5);
        for (TrainInstance train : service.getActiveTrains()) {
            double eta = train.estimateEtaSecondsToStop(startStopId, currentTick,
                    service.getPlugin().getTravelTimeEstimator());
            if (Double.isFinite(eta) && eta <= threshold) {
                return true;
            }
        }
        return false;
    }
}

File: src/main/java/org/cubexmc/metro/service/strategy/GlobalDispatchStrategy.kt
package org.cubexmc.metro.service.strategy

import org.cubexmc.metro.service.DispatchStrategy
import org.cubexmc.metro.service.LineService

class GlobalDispatchStrategy : DispatchStrategy {
    override fun tick(service: LineService, currentTick: Long) {
        if (!service.isDepartureWindow(currentTick)) {
            return
        }
        if (service.isLoopLine && isLoopTrainArrivingSoon(service, currentTick)) {
            return
        }

        service.markDeparture(currentTick)
        service.spawnTrain(currentTick)
    }

    private fun isLoopTrainArrivingSoon(service: LineService, currentTick: Long): Boolean {
        if (!service.isLoopLine) {
            return false
        }
        val line = service.line ?: return false
        val ordered = line.orderedStopIds
        if (ordered.isEmpty()) {
            return false
        }
        val startStopId = ordered[0]
        val threshold = maxOf(2.0, service.headwaySeconds * 0.5)
        for (train in service.activeTrains) {
            val eta = train.estimateEtaSecondsToStop(
                startStopId,
                currentTick,
                service.plugin.travelTimeEstimator,
            )
            if (eta.isFinite() && eta <= threshold) {
                return true
            }
        }
        return false
    }
}
